libraries/lib-time-and-pitch/FormantShifter.cpp

```diff
@@ -0,0 +1,183 @@
+{
+   const auto size = fftSize / 2 + 1;
+   const auto end = std::min(size, size_t(size * factor));
+   std::vector<float> tmp(end);
+   for (size_t i = 0; i < end; ++i)
+   {
+      const int int_pos = i / factor;
+      const float frac_pos = 1.f * i / factor - int_pos;
+      const auto k = MapToPositiveHalfIndex(int_pos, fftSize);
+      const auto l = MapToPositiveHalfIndex(int_pos + 1, fftSize);
+      tmp[i] = (1 - frac_pos) * x[k] + frac_pos * x[l];
+   }
+   std::copy(tmp.begin(), tmp.begin() + end, x);
+   if (end < size)
+      std::fill(x + end, x + size, 0.f);
+   return end;
+}
+} // namespace
+
+FormantShifter::FormantShifter(
+   int sampleRate, double cutoffQuefrency,
+   FormantShifterLoggerInterface& logger)
+    : cutoffQuefrency { cutoffQuefrency }
+    , mSampleRate { sampleRate }
+    , mLogger { logger }
+{
+}
+
+void FormantShifter::Reset(size_t fftSize)
+{
+   mFft = std::make_unique<staffpad::audio::FourierTransform>(fftSize);
+   const auto numBins = fftSize / 2 + 1;
+   mEnvelope.setSize(1, numBins);
+   mCepstrum.setSize(1, fftSize);
+   mEnvelopeReal.resize(numBins);
+   mWeights.resize(numBins);
+}
+
+void FormantShifter::Reset()
+{
+   mFft.reset();
+}
+
+void FormantShifter::Process(
+   const float* powSpec, std::complex<float>* spec, double factor)
+{
+   if (cutoffQuefrency == 0 || !mFft)
+      return;
+
+   const auto fftSize = mFft->getSize();
+   const auto numBins = fftSize / 2 + 1;
+
+   mLogger.Log(fftSize, "fftSize");
+
+   // Take the log of the normalized magnitude. (This assumes that
+   // the window averages to 1.)
+   std::complex<float>* pEnv = mEnvelope.getPtr(0);
+   const float normalizer = GetLog2(fftSize);
+   std::transform(powSpec, powSpec + numBins, pEnv, [&](float power) {
+      // Why -12 ?
+      if (!power)
+         return -12.f;
+      return .5f * GetLog2(power) - normalizer;
+   });
+
+   // Get the cosine transform of the log magnitude, aka the cepstrum.
+   mFft->inverseReal(mEnvelope, mCepstrum);
+   auto pCepst = mCepstrum.getPtr(0);
+   mLogger.Log(pCepst, fftSize, "cepstrum");
+
+   // "Lifter" the cepstrum.
+   const auto binCutoff = int(mSampleRate * cutoffQuefrency * factor);
+   if (binCutoff < fftSize / 2)
+      std::fill(pCepst + binCutoff + 1, pCepst + fftSize - binCutoff, 0.f);
+   mLogger.Log(pCepst, fftSize, "cepstrumLiftered");
+
+   // Get the envelope back.
+   mFft->forwardReal(mCepstrum, mEnvelope);
+   std::transform(
+      pEnv, pEnv + numBins, mEnvelopeReal.begin(),
+      [fftSize = fftSize](const std::complex<float>& env) {
+         return std::exp2(env.real() / fftSize);
+      });
+   mLogger.Log(mEnvelopeReal.data(), numBins, "envelope");
+
+   // Get the weights, which are the ratio of the desired envelope to the
+   // current envelope (which has the effect of downsampling).
+   std::transform(
+      mEnvelopeReal.begin(), mEnvelopeReal.end(), mWeights.begin(),
+      [](float env) { return std::isnormal(env) ? 1 / env : 0.f; });
```

Comment:
As `env` can still be close to infinity small, the resulting weight might get a huge number introducing noise and artifacts. I could imagine, that either limiting `1/env` or the weight would result in less audible artifacts.

Reply:
Agreed, limiting the weights it will be.